src/game/connect4.rs

```diff
@@ -65,20 +89,281 @@ impl GameAdapter for Connect4Adapter<'_> {
+
+            //4.) LeftDomn + RightUp
+            lenl = 0;
+            lenr = 0;
+            row_aux = row;
+            col_aux = column;
+            ok = true;
+            while (lenl < CONNECT_FOUR)
+                && (ok)
+                && (self.get_cell_at(row_aux, col_aux) == Some(self.turn))
+            {
+                lenl += 1;
+                if row_aux > 0 {
+                    row_aux -= 1;
+                } else {
+                    ok = false;
+                }
+                if col_aux > 0 {
+                    col_aux -= 1;
+                } else {
+                    ok = false;
+                }
+            }
+            while (lenr < CONNECT_FOUR)
+                && (col_aux < COL_SIZE)
+                && (row_aux < ROW_SIZE)
+                && (self.get_cell_at(row_aux, col_aux) == Some(self.turn))
+            {
+                lenl += 1;
+                col_aux += 1;
+                row_aux += 1;
+            }
+            lenl + lenr >= CONNECT_FOUR
+        } else {
+            false
+        }
+    }
+    fn draw_move(&self) -> bool {
+        self.board.iter().all(|ref col| col.len() == ROW_SIZE)
+    }
+
+    fn moves(&mut self, column: usize) -> actix_web::Result<()> {
+        if self.completed {
+            return Err(actix_web::Error::from(GameAdapterError::InvalidGameState(
+                State::Ended,
+            )));
+        }
+        self.insert_move_if_legal(column)?;
+        Ok(())
+    }
+    fn visualize(&mut self, column: usize) {
+        let mut ct = 0;
+        let mut tok_turn = "";
+
+        if self.turn == Token::Red {
+            tok_turn = "Red";
+        }
+        if self.turn == Token::Blue {
+            tok_turn = "Blue";
+        }
+        println!("Adding on column {}: on token {}", column, tok_turn);
+        self.insert_move_if_legal(column);
+        println!("Board after inserting:");
+        self.board.iter().for_each(|it| {
+            println!("Col {}, {:#?}", ct, it);
+            ct += 1;
+        });
+        let win = self.winning_move(column);
+        let draw = self.draw_move();
+        println!("Win:{}, Draw:{}", win, draw);
+
+        if win || draw {
+            self.completed = true;
+        } else {
+            self.switch_token();
+        }
+        println!("Completed: {};", self.completed);
+    }
+}
+
+#[cfg(test)]
+mod tests {
+
+    use crate::game::connect4::Connect4;
+    use crate::game::connect4::Token;
+    use crate::game::connect4::COL_SIZE;
+    #[test]
+
+    fn add_() {
+        let mut board1 = Connect4 {
+            completed: false,
+            turn: Token::Red,
+            board: vec![vec![]; COL_SIZE], // vector of columns, each variable length.
+        };
+        board1.visualize(8);
+        board1.visualize(2);
+        board1.visualize(2);
+        board1.visualize(2);
+        board1.visualize(2);
+        board1.visualize(2);
+        board1.visualize(2);
+
+        assert_eq!(2 + 2, 4);
```

Comment:
Should probably use assertions to check that things worked correctly rather than relying on a human to read the output manually